### PR TITLE
Stat Panel: data dependent override stat formatting

### LINKED ISSUE #63528: Stat Panel: Support custom prefixes/suffixes for Stat Panel
This is a feature brought up by Samuel Garen in the Business Intelligence Round Table with the BI Squad from 2023/02/06

### PR DESCRIPTION
**Pull request by baldm0mma**
**What is this feature?**

This feature gives the user the ability to add value-based prefixes and suffixes to numeric stat panel stats.

**Why do we need this feature?**

Highly requested BI feature

**Who is this feature for?**

All users

**Which issue(s) does this PR fix?**:

Fixes #63528

Screenshot 2023-03-17 at 4 00 47 PM (2)

### CODE REVIEW

File: public/app/plugins/panel/stat/override.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+export function applyDisplayOverrides(
+  disp: FieldDisplay,
+  theme: GrafanaTheme2,
+  replace: InterpolateFunction,
+  applyToType: FieldType[]
+): FieldDisplay {
+  const { valueType } = disp.display;
+  // Only apply the display overrides if the valueType (FieldType) has been specified
+  if (applyToType.some((type) => type === valueType)) {
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/override.ts:61]: why do we need this?   you can specify what fields you want the overrides to apply to in normal override path?

File: packages/grafana-data/src/types/displayValue.ts [hunk 2]
```diff
@@ -20,7 +22,10 @@ export interface DisplayValue extends FormattedValue {
    */
   icon?: string;
   title?: string;
-
+  /**
+   *  FieldType for use when adding custom prefixes/suffixes to specific types of values
+   */
+  valueType?: FieldType;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/displayValue.ts:28]: for sure we do not want to add type here... if anything (and i don't think we need anything) lets attach the source field, then we will have it there once and for all :)

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 3]
```diff
@@ -114,7 +114,8 @@ export class StatPanel extends PureComponent<PanelProps<PanelOptions>> {
       data: data.series,
       sparkline: options.graphMode !== BigValueGraphMode.None,
       timeZone,
-    });
+      // Only apply the display overrides if `FieldType.number`
+    }).map((v) => applyDisplayOverrides(v, config.theme2, this.props.replaceVariables, [FieldType.number]));
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:118]: why number?

File: public/app/plugins/panel/stat/override.ts (hunk 1 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/override.ts:61]: Ah, I see now; sorry! Removed.

File: packages/grafana-data/src/types/displayValue.ts (hunk 2 above)
Comment (baldm0mma) [reply packages/grafana-data/src/types/displayValue.ts:28]: Yep; we don't need this anyway.

File: public/app/plugins/panel/stat/StatPanel.tsx (hunk 3 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/StatPanel.tsx:118]: No need! Removed.

**Review by ryantxu — APPROVED** [review]
nice

**Comment by torkelo** [comment]
Give me a chance to review this next week before merging

**Review by torkelo — COMMENTED** [review]
A couple of things here that feel wrong

**Comment by torkelo** [comment]
Will review and test more next week

**Review by torkelo — COMMENTED** [review]
All this PR does can already be accomplished with the standard unit field config option, and the color scheme field config option. 

Screenshot from 2023-03-19 11-27-03

**Comment by baldm0mma** [comment]
@torkelo The feedback we got about the standard units is that they don't support more "custom" options. Some BI users wanted to be able to prepend a `↑` or a `Δ` to values - which the standard units don't support - based on value, as well as do the same with already-applied standard units that have been chosen already: `56%` -> `↑56%`, or `200%` -> `~200%`, etc. Currently it can support color and any standard unit, but not custom prefixes and suffixes, which this was intended to remedy.

**Review by torkelo — CHANGES_REQUESTED** [review]
@baldm0mma I can see that, but why is it adding a new color option, and what is the text option for? 

I would not add the suffix and prefix options either, but something easier to use (that does not require multiple complex overrides)

I imagined solving this with a new "trend" option, which could automatically show the change in value (as percent, or just as arrows).  The thing that has stopped me from implementing it is how to handle the cases when stat panel is given just single values (so no trend) and the trend might be in another field. Supporting that scenario always seemed so hard. But for this scenario of showing up or down arrow or percent change, I think it should be pretty straight forward to implement

**Comment by baldm0mma** [comment]
@torkelo Ah, I see what you mean; I misunderstood before. I think we are indeed, as you mentioned, pretty well covered by the `fields with values` picker and basic overrides. My bad.

**Comment by baldm0mma** [comment]
I think we're good to close this PR now that `fields with values` has been integrated, and custom prefix/suffixing syntax works with the override.

**Pull request closed without merging**